week-four-progress

tanvirastogi08 opened the pull request:
-- Added handlebars to all the view.
-- Able to filter the products on selecting any category.
-- Some code refactoring and code restructuring.
-- Added global config file
-- Used mocked json by creating XMLHttpRequest with promises.

Commit:
Webpack configuration in progress..

tanvirastogi08 committed:
implemented webpack succesfully

tanvirastogi08 committed:
Merge branch 'develop' of https://github.com/tanvirastogi08/shopping-site-coding-exercise into feature/webpack

tanvirastogi08 committed:
Fixed products images path

tanvirastogi08 committed:
Added json file and installed handlebars and handlebars loader

Commit:
Merge branch 'develop' of https://github.com/tanvirastogi08/shopping-site-coding-exercise into feature/webpack

Commit:
Commenting

Commit:
Added dynamic data and implemented it into products, category listing

Commit:
Added Row Reverse to even element of the category on home page

Commit:
Commenting

Commit:
Removed unnecessary folder

tanvirastogi08 committed:
Webpack bundling fixes done

Commit:
Code refactoring

Commit:
using const instead of var

Commit:
Made home page responsive

Commit:
Made product page responsive

Commit:
worked on dynamic labeling of buy now buttons on products page and made banner dots dynamic

Commit:
Added categories dropdown on mobile view of products page using select

Commit:
Added Login/Signup responsiveness

Commit:
some correction

Commit:
Removed unnecessary wrappers and corrected media query

Commit:
Fixed some products page css

Commit:
Margin top in mobile view

Commit:
kept all media query at common place

Commit:
added some shared color variables

tanvirastogi08 committed:
Merged with responsive design and resolved conflicts

tanvirastogi08 committed:
able to test handlebar integration

Commit:
Added handlebars for home,products,signup and login page

Commit:
some correction

tanvirastogi08 committed:
Promise handling for mock data and some code refactoring

tanvirastogi08 committed:
changed console.log to console.error

tanvirastogi08 committed:
folder restructuring

tanvirastogi08 committed:
Able to filter products by category

tanvirastogi08 committed:
Able to select the products by category on mobile view as well

tanvirastogi08 committed:
updated README.md file

Commit:
Tested web accessiblity and added tabIndex wherever needed

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+        <span tabindex="0">Copyright @ 2011-2019 Sabka Bazar Grocery Supplies Pvt Ltd</span>
       </div>
     </footer>
+    <script>
```
Why you keeping inline script and not putting it in corresponding module.

tanvirastogi08 replied:
Updated.

vinodloha commented on config/mock-data/addToCart/index.post.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Does all of API calls have same structure of response? Usually its better to do it consistently 

`{
status: "success",
data:{
  "message": "Product added to cart successfully"
  }
}`

tanvirastogi08 replied:
Not in use, will update the structure with all the consistencies ASAP.

vinodloha commented on config/mock-data/banners/banners.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "banners": [
+    {
+      "bannerImageUrl": "/images/offers/offer1.jpg",
```
How are you doing responsive images with these type of image paths?

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+
     <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Dosis:400,600,700">
-    <link type="text/css" rel="stylesheet" href="./shared/css/style.css">
     <script src="http://code.jquery.com/jquery-3.3.1.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
```
Why is this Script in the Head section?

tanvirastogi08 replied:
Updated.

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
     <title>Sabka Bazar</title>
-    <!-- <base href="/"> -->
+
     <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Dosis:400,600,700">
```
For now its okay, but this isn't the only way to use web fonts. There are better and different approaches available. Please explore them as well. Read about https://css-tricks.com/fout-foit-foft/

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
   <body>
-    <header class="header">
+    <header>
       <nav class="navigation">
```
If you are using HBS why this markup is part of your HTML page, it should be inside HBS components and at page (template) level you only import those components.

tanvirastogi08 replied:
Created a separate component for both Header and Footer.

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
-          <li><a href="#" onclick="onNavItemClick('home')">Home</a></li>
-          <li><a href="#" onclick="onNavItemClick('products')">Products</a></li>
-        </ul>
+        <img src="images/logo.png" alt="Sabka Bazar logo" class="logo">
```
How will this logo look at Retina device? Why are we using the same image for all view-ports? Go through this https://responsiveimages.org/

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+        <div class="drawer-nav" id="drawer">
+          <a href="#" class="cross-icon" onclick="onDrawerCloseIconClick()">&times;</a>
+          <a href="#" class="nav-home">Home</a>
+          <a href="#" class="nav-product">Products</a>
```
Why some of the NAV element are coming twice? such duplication should be avoided.

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+            <span class="items-in-cart">0</span>
+            <span class="items-text">&nbsp;items</span>
+          </a>
+          <span class="right-nav" onclick="onDrawerIconClick()">&#9776;</span>
```
Don't write JS in HTML, learn about unobtrusive JS

tanvirastogi08 replied:
Updated.

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+          <a href="#" class="btn-register">SignUp</a>
         </div>
       </nav>
     </header>
```
How does a disabled user jump to main content? They will have to go thorough each tab!

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+    <!-- Keeping constants on global window -->
+    <script>
+      // https://github.com/auth0/auth0.js/issues/753
+      window.global = window;
```
WHAT? WHY? We ate not using angular so why do you need this type of hacks?

tanvirastogi08 replied:
Sorry for this. Updated.

vinodloha commented on index.html:
```diff
@@ -4,44 +4,69 @@
+    </script>
+
+    <!-- Keeping constants on global window -->
+    <script>
```
Why so many script tags, do you know they are render blocking? https://developers.google.com/speed/docs/insights/BlockingJS

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/core/services/sabkabazar.service.js:
```diff
@@ -0,0 +1,36 @@
+export function makeRequest(url, method) {
```
Why this module is so poorly formatted? Use  https://prettier.io/

vinodloha commented on src/app/core/services/sabkabazar.service.js:
```diff
@@ -0,0 +1,36 @@
+export function makeRequest(url, method) {
+
+	// Create the XHR request
+	var request = new XMLHttpRequest();
```
Why not use https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

tanvirastogi08 commented:
Hi Vinod,

I will look into all the comments to make improvements and will explain to you the things wherever needed. Thanks for reviewing.

vinodloha commented on src/app/home/home.hbs:
```diff
@@ -0,0 +1,33 @@
+    <div class="slideshow-container">
+        {{#each banners.banners}}
+        <div class="bannerSlides fade">
+                <img src="{{bannerImageUrl}}" style="width:100%" alt="{{bannerImageAlt}}">
```
Again, no responsive image

vinodloha commented on src/app/home/home.hbs:
```diff
@@ -0,0 +1,33 @@
+                <img src="{{bannerImageUrl}}" style="width:100%" alt="{{bannerImageAlt}}">
+        </div>
+        {{/each}}
+      <a class="prev">&#10094;</a>
```
Anchor without Href is accessibility's P1 defect

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/home/home.hbs:
```diff
@@ -0,0 +1,33 @@
+      <a class="prev">&#10094;</a>
+      <a class="next">&#10095;</a>
+      <br>
+      <div style="text-align:center;">
```
Don't use inline CSS

tanvirastogi08 replied:
Updated. Will keep this in mind in future.

vinodloha commented on src/app/home/home.hbs:
```diff
@@ -0,0 +1,33 @@
+      <br>
+      <div style="text-align:center;">
+        {{#each banners.banners}}
+        <span class="dot dot-{{@index}}"></span>
```
Its clickable element, why its a <span> and not a button??

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/home/home.hbs:
```diff
@@ -0,0 +1,33 @@
+                <div class="product-detail">
+                    <h5>{{name}}</h5>
+                    <p>{{description}}</p>
+                    <a class="btn-explore-category" href="#">Explore-{{key}}</a>
```
href="#" should be driven with a dynamic value too

vinodloha commented on src/app/home/home.hbs:
```diff
@@ -0,0 +1,33 @@
+    <div class="products-section">
+      <ul class="products-listing">
+            {{#each categories.categories}}
+            <li class="category-list">
```
Why this product is not an independent component?

vinodloha commented on src/app/home/home.js:
```diff
@@ -1,120 +1 @@
-function currentSlide(n) {
-  this.showSlides(this.slideIndex = n);
-}
+export const home = require('./home.hbs');
```
Why you not using import and loading with require('./home.hbs')

vinodloha commented on src/app/home/home.scss:
```diff
@@ -1,12 +1,13 @@
 .home-container {
-  margin: 0 8rem;
+  margin: 0 7rem;
```
Usually we set `margin: 0 auto;` on such containers instead of fixed gutters

vinodloha commented on src/app/home/home.scss:
```diff
@@ -1,12 +1,13 @@
+.bannerSlides {display: none}
 
-/* Slideshow container */
+img { vertical-align: middle; }
```
Where did you handle Image Scaling? https://www.geeksforgeeks.org/resize-image-proportionally-with-css/

vinodloha commented on src/app/home/home.scss:
```diff
@@ -1,12 +1,13 @@
+
+/************ Slideshow container *************/
 .slideshow-container {
   max-width: 1000px;
```
How do you know 1000px would be enough?

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/home/home.scss:
```diff
@@ -102,25 +82,39 @@ img {vertical-align: middle;}
+    padding: 10px 0;
+    box-shadow: 0 2px 2px $color-grey-light;
+
+    .product-category {
```
Why you are not following BEM or SMACSS methodology?

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+import { home } from '../app/home/home';
+import { products, onClickCategory } from '../app/products/products';
+import { login } from '../app/login/login';
+import { signup } from '../app/signup/signup';
```
Instead of repeating the package name twice like ../app/**signup/signup** why don't you place an index.js file in ../app/signup and return the component/module from there?

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  makeRequest(`${WINDOW_CONFIG.apiUrl}/mock-data/banners/banners.json`, 'GET')
+    .then(function(bannersData) {
+
+      makeRequest(`${WINDOW_CONFIG.apiUrl}/mock-data/categories/categories.json`, 'GET')
```
Wrong!! Why these 2 request can't go in parallel?

tanvirastogi08 replied:
Fixed it.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  let handlebarsData;
+
+  // fetch banners and then categories
+  makeRequest(`${WINDOW_CONFIG.apiUrl}/mock-data/banners/banners.json`, 'GET')
```
This code is not organized correctly. There should have an independent module handling this bootstrapping code; here you could have just invoked it.

tanvirastogi08 committed:
1. Created seperate header and footer components
2. Moved the script to respective module.
3. Made clickable elements as buttons instead of Span.
4. Created parallel promises for banner and category fetching.

tanvirastogi08 committed:
Resolved merge conflicts

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+        });
+    })
+    .catch(function (error) {
+      console.error('Something went wrong in banners', error);
```
Shouldn't this type of error handling be happening at one common place?

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+  /**================== Navigation buttons ============================ */
+
+  $('.nav-home').on('click', function() {
```
Why are you not delegating the Events. Learn here why its better and how you can do it https://learn.jquery.com/events/event-delegation/

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  /**================== Navigation buttons ============================ */
+
+  $('.nav-home').on('click', function() {
+    handlebarsData = {
```
What is the point of creating this new variable when you only use it once, 
also why you are not using const/let or var when declaring it, Do you know it makes it global?

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+$(document).ready(function() {
+
+  let handlebarsData;
```
Why is this variable defined in this scope??

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+      })
+  });
+
+  $('.btn-signin').on('click', function() {
```
Both `$('.btn-signin').on('click',` and `$('.btn-register').on('click',` do almost the same thing, you should find ways to avoid such duplication in the code.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+  /**================== Login/Signup focus ============================ */
+  $('.input100').each(function(){
+    $(this).on('blur', function(){
```
What is the need for attaching events through the loop? Read and follow event delegation comment above

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  /**================== Login/Signup focus ============================ */
+  $('.input100').each(function(){
+    $(this).on('blur', function(){
+      if($(this).val().trim() != "") {
```
This IF/Else block is sub-optimal code, see how $.toggleClass() can help you

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+    })    
+  });
+
+  let showPass = 0;
```
Why you doing such hacky solutions? Even if you need it use Boolean for flags. Event attachment and non delegation is global problem

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  
+});
+
+function onNavItemClick(section, data) {
```
Why you are not writing ES6/7/8 format of code?

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+// appending the dynamic html to the main element
+function createHTML(section, data) {
+  content.innerHTML =  routes[section](data);
```
Where does this content come from??

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+function onInitSlider() {
+  // prev and next slider button
+  $('.prev').on('click', function() {
+    plusSlide(-1);
```
Why the function name is plusSlide and not changeSlide??

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+    plusSlide(1);
+  });
+
+  const bannerList = banners.banners;
```
banners.banners ??? why such structure?

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+  const bannerList = banners.banners;
+  bannerList.forEach(function(banner, index) {
+    $(`.dot-${index}`).on('click', function() {
```
Again event attachment in loop is wrong, its absolutely unnecessary; do delegation instead.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  const allCategories = categories.categories;
+
+  // on mobile screen select is visible
+  $('select').on('change', function() {
```
Don't attach event on exact element, delegate it

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+  // on tablet and desktop screen
+  allCategories.forEach((category, index) =>  {
+    $(`.item-${index}`).on('click', function(el) {
```
Use delegation way of attaching an event, there will not be any need for loops

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+}
+
+function onCategorySelection(selectedCategory) {
+  const allCategories = categories.categories;
```
Why we have this kind of data structure categories.categories, it should be categories.items/ categories.list or just categories.

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+
+function onCategorySelection(selectedCategory) {
+  const allCategories = categories.categories;
+  const selectedCategoryDetail = allCategories.filter(item => item.name === selectedCategory);
```
Why can't you figure the category out from DOM element, these are unnecessary loops you are running

vinodloha commented on src/app/index.js:
```diff
@@ -1,69 +1,208 @@
+  const handlebarsData = {
+    products, categories
+  }
+  onClickProductOrCategory('products', handlebarsData)
```
When you are just passing the data down, why we need this variable handlebarsData ??

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/login/login.hbs:
```diff
@@ -0,0 +1,17 @@
+    <form class="login-form">
+        <div class="wrap-input100">
+            <input class="input100" type="text" name="email">
+            <span class="focus-input100" data-placeholder="Email"></span>
```
Why no <label> tags? How will you achieve accessibility?

vinodloha commented on src/app/login/login.hbs:
```diff
@@ -0,0 +1,17 @@
+        <h3>Login</h3>
+        <p style="font-size: 80%;">Get access to your Orders, Wishlist and Recommendations</p>
+    </div>
+    <form class="login-form">
```
How will you will ensure security of user data? This form doesn't have even basic best practices backed in.

vinodloha commented on src/app/login/login.hbs:
```diff
@@ -0,0 +1,17 @@
+<div class="login-container">
+    <div class="login-text-box">
+        <h3>Login</h3>
+        <p style="font-size: 80%;">Get access to your Orders, Wishlist and Recommendations</p>
```
No inline style

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/login/login.hbs:
```diff
@@ -0,0 +1,17 @@
+    </div>
+    <form class="login-form">
+        <div class="wrap-input100">
+            <input class="input100" type="text" name="email">
```
If the field will receive email, why the type doesn't reflect that? Is input type="email" good and helpful on mobile?

vinodloha commented on src/app/login/login.hbs:
```diff
@@ -0,0 +1,17 @@
+            <input class="input100" type="password" name="pass">
+            <span class="focus-input100" data-placeholder="Password"></span>
+        </div>
+        <a class="btn-login" href="#" onclick="clickMe()">Login</a>
```
An <a> tag for submit button ??? are you kidding me?? Where did you learn that?

tanvirastogi08 replied:
It wasn't intentional! Apologies for doing such a thing.

vinodloha commented on src/app/login/login.hbs:
```diff
@@ -0,0 +1,17 @@
+            <input class="input100" type="password" name="pass">
+            <span class="focus-input100" data-placeholder="Password"></span>
+        </div>
+        <a class="btn-login" href="#" onclick="clickMe()">Login</a>
```
onclick="clickMe()" Why you writing Obtrusive markup?

tanvirastogi08 replied:
It was just for some testing purpose. Removed that now.

tanvirastogi08 committed:
1. Kept jquery as an asset

vinodloha commented on src/app/login/login.scss:
```diff
@@ -13,20 +13,19 @@ input:focus {
   display: flex;
   justify-content: space-around;
   margin: 6rem 12rem;
+
```
Make component specific CSS modules an not page specific

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+    <ul class="navbar">
+      {{#each categories.categories}}
+        <li class="nav-item">
+          <a class="nav-item-link item-{{@index}}" href="#">
```
Why you will have an <a> for product with no href?

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+    </ul>
+  </nav>
+  <nav class="topbar">
+    <select class="navbar selected">
```
Why a select inside Nav?

tanvirastogi08 replied:
Because it is used with the role of navigation. So, I thought should wrap it into nav element tag.

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+    <select class="navbar selected">
+      {{#each categories.categories}}
+        <option class="nav-item">
+          <span class="nav-item-link">{{name}}</span>
```
Span inside Option??? it only accepts text. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option

tanvirastogi08 replied:
Updated.

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+      {{/each}}
+    </select>
+  </nav>
+  <ul class="products-showcase">
```
Why its not a separate component?

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+      <h4 tabindex="0">{{name}}</h4>
+      <div class="product-box">
+        <figure class="product-photo">
+          <img tabindex="0" src="{{imageURL}}" alt="{{name}}">
```
Responsive image support?

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+        <figure class="product-photo">
+          <img tabindex="0" src="{{imageURL}}" alt="{{name}}">
+        </figure>
+        <p tabindex="0">{{description}}</p>
```
why you suing tab index so much, and on P it doesn't make any sense

tanvirastogi08 replied:
Tab Index because want the user to read out the description of the products as well.

vinodloha commented on src/app/products/products.hbs:
```diff
@@ -0,0 +1,39 @@
+        <p tabindex="0">{{description}}</p>
+      </div>
+      <div class="mrp-buy-wrapper">
+        <span>MRP Rs. {{price}}</span>
```
Look non semantic markup, why you guys use span so much?

vinodloha commented on src/app/signup/signup.hbs:
```diff
@@ -0,0 +1,29 @@
+        <h3>Signup</h3>
+        <p style="font-size: 80%;">We do not share your personal details with anyone.</p>
+    </div>
+    <form class="signup-form">
```
Same feedback as login form

vinodloha reviewed (changes requested):
There are lot of improvement scope, best practices are completely missing!!

tanvirastogi08 committed:
1. Indentation corection in service.
2. Removed script from head section.
3. Corrected the json data structure
4. Removed inline CSS.
5. Created button wherever needed instead of anchor.
6. Removed unnecessary handlebar variable.
7. Renamed the components variables to make it meaningful(eg: home to HomeComponent)

Commit:
Added basic Web Accessibilty to the home page

tanvirastogi08 committed:
Renamed plusSlides to changeSlides

Commit:
1. Removed SPAN from select-option 2.Removed unnecessary loop for fetching category id 3.Removed next and prev button from Slider on mobile view

Commit:
Toggling hamburger menu, and aria label on carousal dots

tanvirastogi08 committed:
Managed routing, coverted everything to ES6, Removed Jquery as not required for now etc

tanvirastogi08 committed:
Made logo clickable and changed the structuring of the header items and made signin and register first in the row for accessibility

tanvirastogi08 committed:
Able to search the products by category id

tanvirastogi08 committed:
Fixed the categories listing view in mobile

Commit:
Added redirection to products category on click on home page and empty cart modal popup

Commit:
fixed cart popup close

Commit:
Added formaction in login button

tanvirastogi08 committed:
Added Error Page
